### PR TITLE
Added factory method to GeoLocationInformation to create instance with null checks

### LINKED ISSUE #12788: Null pointer exceptions in GeoIP processor after update to 4.3
## Current Behavior
Following an upgrade to Graylog 4.3.1, my server.log in now flooded with the following Warnings :

```
2022-06-01T23:23:58.280+01:00 WARN  [ProcessBufferProcessor] Unable to process message <87c84c80-e1f9-11ec-8ea4-005056afb272>: java.lang.NullPointerException
2022-06-01T23:23:58.666+01:00 WARN  [ProcessBufferProcessor] Unable to process message <88030b91-e1f9-11ec-8ea4-005056afb272>: java.lang.NullPointerException
2022-06-01T23:23:58.667+01:00 WARN  [ProcessBufferProcessor] Unable to process message <880359b1-e1f9-11ec-8ea4-005056afb272>: java.lang.NullPointerException
2022-06-01T23:23:59.761+01:00 WARN  [ProcessBufferProcessor] Unable to process message <88aa2107-e1f9-11ec-8ea4-005056afb272>: java.lang.NullPointerException
2022-06-01T23:23:59.769+01:00 WARN  [ProcessBufferProcessor] Unable to process message <88ab8090-e1f9-11ec-8ea4-005056afb272>: java.lang.NullPointerException
2022-06-01T23:23:59.778+01:00 WARN  [ProcessBufferProcessor] Unable to process message <88acb916-e1f9-11ec-8ea4-005056afb272>: java.lang.NullPointerException
2022-06-01T23:24:00.390+01:00 WARN  [ProcessBufferProcessor] Unable to process message <890a1b66-e1f9-11ec-8ea4-005056afb272>: java.lang.NullPointerException
2022-06-01T23:24:00.682+01:00 WARN  [ProcessBufferProcessor] Unable to process message <8936a992-e1f9-11ec-8ea4-005056afb272>: java.lang.NullPointerException
2022-06-01T23:24:00.949+01:00 WARN  [ProcessBufferProcessor] Unable to process message <895f6741-e1f9-11ec-8ea4-005056afb272>: java.lang.NullPointerException
2022-06-01T23:24:01.286+01:00 WARN  [ProcessBufferProcessor] Unable to process message <8992d351-e1f9-11ec-8ea4-005056afb272>: java.lang.NullPointerException
2022-06-01T23:24:01.326+01:00 WARN  [ProcessBufferProcessor] Unable to process message <8998eddd-e1f9-11ec-8ea4-005056afb272>: java.lang.NullPointerException
2022-06-01T23:24:01.727+01:00 WARN  [ProcessBufferProcessor] Unable to process message <89d61de5-e1f9-11ec-8ea4-005056afb272>: java.lang.NullPointerException
2022-06-01T23:24:02.058+01:00 WARN  [ProcessBufferProcessor] Unable to process message <8a08c6a3-e1f9-11ec-8ea4-005056afb272>: java.lang.NullPointerException
2022-06-01T23:24:03.330+01:00 WARN  [ProcessBufferProcessor] Unable to process message <8acade21-e1f9-11ec-8ea4-005056afb272>: java.lang.NullPointerExceptio
```

What can I do to understand what is causing these warnings and how to stop it ?

## Your Environment

* Graylog Version: 4.3
* Java Version: 8u312
* Elasticsearch Version: 7.10.2
* MongoDB Version: 4.0.28
* Operating System: Ubuntu Server 20.04

### PR DESCRIPTION
**Pull request by roberto-graylog**
Updated MaxMindIpLocationResolver to use null-safe method to create GeoLocationInformation instance.
closes #12788 

## Description

Addresses a NullPointerException error caused in MaxMind Geo IP Information loader when the returned object contains null longitude/latitude fields.

To address the issue, a new method was added to GeoLocationInformation to create an instance from a Location, Country, and City object.  All the parameters as well as the needed fields from each parameter are checked for null.

## Motivation and Context

The error prevents the entire message from being processed.

## How Has This Been Tested?

A new unit test was added to test that no exception is thrown if the MaxMind database reader returns null city response fields, or if any of the required response fields themselves have null fields.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

**Comment by bernd** [comment]
Please create a 4.3 backport once the PR has been reviewed and merged. Thank you!

File: graylog2-server/src/main/java/org/graylog/plugins/map/geoip/MaxMindIpLocationResolver.java [hunk 1]
```diff
@@ -46,28 +46,23 @@ public MaxMindIpLocationResolver(@Assisted Timer resolveTime,
...
-            info = null;
+            GeoLocationInformation gli = GeoLocationInformation.create(location, country, city, "N/A");
+            info = Optional.of(gli);
+        } catch (IOException | GeoIp2Exception | NullPointerException | UnsupportedOperationException e) {
+            info = Optional.empty();
             if (!(e instanceof AddressNotFoundException)) {
                 LOG.debug("Could not get location from IP {}", address.getHostAddress(), e);
                 lastError = e.getMessage();
             }
         }
```
Comment (danotorrey) [inline graylog2-server/src/main/java/org/graylog/plugins/map/geoip/MaxMindIpLocationResolver.java:64]: Just one small suggestion: I think the `Optional<GeoLocationInformation> info` variable declaration on line 49 can be avoided by returning the new optionals in-line:
```
        try (Timer.Context ignored = resolveTime.time()) {
            final CityResponse response = databaseReader.city(address);
            final Location location = response.getLocation();
            final Country country = response.getCountry();
            final City city = response.getCity();

            return Optional.of(GeoLocationInformation.create(location, country, city, "N/A"));
        } catch (IOException | GeoIp2Exception | NullPointerException | UnsupportedOperationException e) {
            if (!(e instanceof AddressNotFoundException)) {
                LOG.debug("Could not get location from IP {}", address.getHostAddress(), e);
                lastError = e.getMessage();
            }
            return Optional.empty();
        }
```

**Review by danotorrey — APPROVED** [review]
This LGTM! Thanks a lot for the additional unit tests. 

I just left one small suggestion inline.

I am still looking into how to test it end-to-end. No luck so far finding an IP address that does not have location info. Have you been able to find one?

I was digging through the MaxMind source code and found this in `com.maxmind.geoip2.model.AbstractCityResponse`, which makes me wonder how the location ends up being null.
`this.location = location != null ? location : new Location();`

I will continue digging on how to end-to-end test...

File: graylog2-server/src/main/java/org/graylog/plugins/map/geoip/MaxMindIpLocationResolver.java (hunk 1 above)
Comment (roberto-graylog) [reply graylog2-server/src/main/java/org/graylog/plugins/map/geoip/MaxMindIpLocationResolver.java]: Updated in the latest push.  I should note that having a single exit point from a function/method is the preferred approach/best practice.  

Personally I find method calls as method arguments such as `Optional.of(GeoLocationInformation.create(location, country, city, "N/A"))` difficult to read, in particular when the second (or any number that follows) has more than one argument.

**Review by ryan-carroll-graylog — APPROVED** [review]
Looks good to me

**Pull request merged**